### PR TITLE
Jack devel remove rosbuild

### PR DESCRIPTION
**Pull request by JJBUT**
Roughly same thing as amrl_msgs. Again the only frustration is setting the CMAKE_PREFIX_PATH for amrl_msgs.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -18,17 +18,19 @@ ENDIF()
...
-INCLUDE($ENV{ROS_ROOT}/core/rosbuild/rosbuild.cmake)
-rosbuild_init()
-SET(ROS_BUILD_STATIC_LIBS true)
-SET(ROS_BUILD_SHARED_LIBS false)
-
 SET(EXECUTABLE_OUTPUT_PATH ${PROJECT_SOURCE_DIR}/bin)
 SET(LIBRARY_OUTPUT_PATH ${PROJECT_SOURCE_DIR}/lib)
 
+# Path to amrl message's "amrl_msgsConfig.cmake"
+SET(CMAKE_PREFIX_PATH "/home/administrator/PROJECTS/fault_monitoring/amrl_msgs/build/devel/share/amrl_msgs/cmake")
```
Comment (joydeep-b) [inline CMakeLists.txt:25]: Sorry, this won't fly. We cannot hard-code paths in CMakeLists.

**Review by joydeep-b — CHANGES_REQUESTED** [review]
We cannot use a solution that involves hardcoding paths. If that is the only way to do this, sorry we will not do it.

**Comment by JJBUT** [comment]
Exporting ROS_PACKAGE_PATH is  the same as hard coding paths imho, but a little easier for people to do, but I understand this would be untenable as the number of packages requiring hard coded paths increases, but again so will the number of times you need to export ROS_PACKAGE_PATH. 

I don't really like the solution I made here, but I wanted to get it in front of more eyes, cause rosbuild is going to be a frustration eventually.

Do you know of a way to make the amrl_msgsConfig discoverable, maybe by installing it in a more generic place meant for that type of config file?

If not ill toss these branches

**Comment by joydeep-b** [comment]
Exporting ROS_PACKAGE_PATH is not the same as hard coding paths - the
former is not under version control, and each user can customize to their
layout. Hardcoding paths means that it's in the repo, and everyone needs to
have the same layout.

**Comment by JJBUT** [comment]
Point taken. I understand hard coded paths is a fundamentally bad thing and was thinking about the process of just getting a package to build, not the implications of version control.

"Do you know of a way to make the amrl_msgsConfig discoverable, maybe by installing it in a more generic place meant for that type of config file?"

**Comment by joydeep-b** [comment]
The "usual" solution to making CMake package finding scripts discoverable is to use the `install` command to copy them to system-wide locations. 
Not sure we want to go that route - it adds an additional step (installing the packages that need to be discovered by others), making it more onerous rather than making it easier. And if the packages change, the updates need to be re-installed.

**Comment by JJBUT** [comment]
Ok, thanks for the information. Well I searched pretty far and wide, and the install route seems to be the only viable option, as there is none built into ROS or any other part of the ecosystem I could discover. 

Looks like rosbuild is here to stay, but if we do want to think about another route im happy to test it out.

**Comment by kvsikand** [comment]
Another note on the install route is that it will mean that if anyone wants to, for example, use one of our packages on a shared computer on which they do not have `sudo` access, they will not be able to do so. (There is the user-space install thing, but as I recall it's pretty tough to get working correctly)

**Comment by JJBUT** [comment]
Ok this is the proper way to do it, essentially just use CMAKE_PREFIX_PATH instead of ROS_PACKAGE_PATH. See the following commit I made on my amrl_msgs branch to the makefile and readme. Is this a good solution?

[commit](https://github.com/ut-amrl/amrl_msgs/pull/1/commits/9740a846757dea0bf62a0dc8733d2e4f89c43807)

**Comment by joydeep-b** [comment]
@JJBUT I read through this PR and the related amrl_msgs PRs. It looks like this change will add another step to the build process (having to manually export the cmake prefix path env variable) for all dependent packages of amrl_msgs, and other dependencies that get this treatment. Given that it makes the process more complex rather than less, can you explain again what the benefit of this change is?

**Comment by JJBUT** [comment]
@joydeep-b thanks for reviewing this!

Is it a requirement that you have to manually export the ROS_PACKAGE_PATH env variable for every package in a similar way? I don't use rosbuild a ton so I may be misunderstanding this requirement or some novel aspect of it. And as I understand it, once you install a package like amrl_msgs you put the export line in your .bashrc and just forget about it, and all future dependent packages will be able to locate it. 

This way instead of exporting your ROS_PACKAGE_PATH in your .bashrc you export the CMAKE_PREFIX_PATH. The benefit would be to use a more "pure" build system like you advocate as the reason for not using catkin, and removing a dependency on rosbuild which has been discouraged for years.

If in simple terms you can explain the "novel aspect of it" that I am missing, I can finally put this horse to bed and stop bugging you guys lol. Thanks!

**Comment by joydeep-b** [comment]
Here are the three ideal requirements for a build system:
1. Compilation of packages should require only user-space edits
2. It should be completely agnostic of where repos are kept, and it should be up to the user where and how they organize their projects.
3. There should be no env. variables or other out of band requirements.

CMake system installs rule out 1, Neither ROS nor CMake can handle 3, and 2 rules out catkin. Hence, we are left with rosbuild, but now the modified requirements are:
1. Compilation of packages should require only user-space edits
2. It should be completely agnostic of where repos are kept, and it should be up to the user where and how they organize their projects.
3. There should be **minimal** env. variables or other out of band requirements.

The edit to ROS_PACKAGE_PATH is the only env. variable that we begrudgingly have to make now. We don't want to have to add yet another env. variable to edit. 

With a large number of projects, editing env variables becomes tedious - I have ~40 lines of ROS_PACKAGE_PATH edits in my bashrc, and don't want to have to double (and maintain! :scream:  ) that for another env. var per package.

**Comment by JJBUT** [comment]
@joydeep-b I don't think I explained myself clearly, sorry about that! 

But the ONLY thing you need to do is export the CMAKE_PREFIX_PATH. So in my testing, this allows me to build the package very simply, and still exporting exactly one env variable as was the case before, but now not requiring rosbuild.

**Comment by joydeep-b** [comment]
Let's discuss this in the lab meeting today.

**Pull request closed without merging**